### DISCUSSION TITLE
Proposal: Python like block indentation

### DISCUSSION THREAD

**Discussion by Developer-Alexander** [dotnet/csharplang#6360] [Language Ideas]
C# is desinged as as a curly bracket based language in that code blocks are defined by encapsulating them in curly brackets.
In comparison, in Python blocks are defined or separated from each other by appropriate indentation. This can result in the code being more compact and clearer, and therefore easier to read and understand. This might be considered by some people as a major advantage of Python.

For this reason, I would like to propose the possibility of defining blocks of code using indentation instead of curly braces in C# as well.
To avoid contradictions and ambiguities, indentation-based blocks could only be allowed in special areas e.g. in between `#indentation_based_blocks enable` and `#indentation_based_blocks disable`.

Here is an example with some dummy code for illustration purposes:

- Classic C# code

```
public class Test
{
    static int Fibonacci(int x)
    {
        if (x == 0)
        {
            return 0;
        }

        int prev = 0;
        int next = 1;
        for (int i = 1; i < x; i++)
        {
            int sum = prev + next;
            prev = next;
            next = sum;
        }
        return next;
    }

    public static void Main()
    {
        Console.Write(Fibonacci(10));
    }
}
```

-Code with indentation based blocks

```
#indentation_based_blocks enable
public class Test
    static int Fibonacci(int x)
        if (x == 0) 
            return 0;

        int prev = 0;
        int next = 1;
        for (int i = 1; i < x; i++)
            int sum = prev + next;
            prev = next;
            next = sum;
        return next;

    public static void Main()
        Console.Write(Fibonacci(10));
# indentation_based_blocks disable
```

In case of indentation based blocks the trailing semicolon might become optional in some cases. 

On the other hand a mechanism is will be needed to indicate that when statement will continue on the next line. (like `\` in Python)

**Comment by DavidArno** [comment]
You appear to be asking for the ability to write C# along these lines:
```f#
let fibonacci n = 
    let mutable last = 0
    let mutable next = 1
    seq {
        0
        1
        for i in 1 .. (n - 1) do
            let temp = last + next
            last <- next
            next <- temp
            next
        }
```        

Good news, there's already a way of doing this: just use F# instead. 😇

**Reply by Developer-Alexander** [reply]
F# is not an attractive option just because it has this one feature.

I am basically asking to save some curly brackets.

**Comment by HaloFour** [comment]
C# is not a whitespace sensitive language and there is no desire to make it one.  Adding flags effectively create dialects of the language which is something the team is extremely loathe to do, and to define what is already existing syntax to mean something else would only cause confusion.

To David's point, if you prefer the style of whitespace sensitive languages there are several to choose from.  This is not a feature that would ever be considered for C#, just as braces would never be considered to drive scoping in Python.

**Reply by Developer-Alexander** [reply]
I agree that creating dialects of a language is something that you have to be careful with (in general). But every version of the C# language is a dialect that is not compatible to previous ones. (C# 10 is one, C# 9 is a dialect and so on...) If there should be no new dialects of C#, then you could actually close this repo for new language features.

When blocks are defined based on identation it would be an side effect that the language is somehow whitespace sensitive. But is this a problem? It's not like that the majority of people choose a programming language because it is not whitespace sensitive. (my assumption) There are many reason why people choose a programming language of course. But I asume that getting the job done with less typing while having readable understandable code is a major one. While C# already being an awesome language I think my proposal is something with that C# could become even better in this aspect.

Of course we don't know the numbers, but don't think there might be a lot of developers who might like this proposal. Especially considering Python's popularity. Maybe it will also lead to one or the other switching to the .NET ecosystem.

**Reply by 333fred** [reply]
> But every version of the C# language is a dialect that is not compatible to previous ones. (C# 10 is one, C# 9 is a dialect and so on...)

Sure, and with the current flags we have, each new language version is currently 4 versions (nullable on/off, checked onn/off). Add another new flag, and suddenly each version is _8_ new versions. This grows exponentially, and that's leaving aside the fact that we are not going to make this change.

**Reply by 333fred** [reply]
> But I asume that getting the job done with less typing while having readable understandable code is a major one.

And this proposal would remove a lot of shorthand that developers use today for short syntax. Expression-bodied members, for example, now gone. Lambdas now must be multiple lines (and where would the indentation even go? Indented from the containing statement? Indented from the method header?). Indentation-based languages are not a universal win wrt code size, and we have no desire to make such a big change to C# that would remove a lot of our existing shorthand.

**Reply by CyrusNajmabadi** [reply]
To add to what Fred said, this would also be *extremely* expensive to implement. We would likely need years of work on the ide to get all our features to work well with whitespace indentation.  Unlike some features which can be pretty cheaply added (like one-two people for a few weeks), this would need the entire team on nothing but this for a very long time.

**Reply by Xyncgas** [reply]
On the other hand, if we do this we can do a bunch other thing since we are creating different flavors of the language
using bracket has it's own beauty, I for one just think about brackets on my bed when I was young
if you are writing codes in c# somtimes you don't have to use brackets, turn your code into procedural style, write oneliner directly under if statements

**Reply by Developer-Alexander** [reply]
> We would likely need years of work on the ide to get all our features to work well with whitespace indentation.

Don't you think that this can be handled at the beginning of the compilation process when the SyntaxTree is created?

**Reply by CyrusNajmabadi** [reply]
No. It would affect each feature.

**Comment by 333fred** [comment]
See https://github.com/dotnet/csharplang/issues/1581. But tldr: no. C# will not be changing to be a whitespace sensitive language.

**Comment by Developer-Alexander** [comment]
It looks like my proposal has no chance of success. 

But thank you anyway.

**Comment by huoyaoyuan** [comment]
> This can result in the code being more compact and clearer, and therefore easier to read and understand

I'd really argue about this. Lack of explicit *end of scope* token can be easily error-prone when modifying code. I've experienced this when I'm writing Python and I've always treated Python as an opposite sample of language design.

There are styles from other languages that I do support, for example type in the following. But it would be extremely difficult to change for any mature language.

**Comment by TahirAhmadov** [comment]
2 thoughts on the topic: 1) this is another good candidate for the "FAQ" - no, C# will not become an indentation-based language, and 2) popularity of Python among beginners and specialists in adjacent fields, like physics and such, creates this phenomena where people start asking questions like, "why isn't your language like Python?" - it became "native tongue" for too many; a good reminder to reinforce efforts to make C# an attractive option to those users.

**Reply by HaloFour** [reply]
I think that's true of syntax questions in general.  C# isn't going to start to look like another programming language for the sake of looking like another programming language, regardless of how much individuals may prefer the style of those programming languages.  Individual features from other languages may be considered, but they would be adapted to fit the existing syntax and style of C#.

I think it's safe to say that C# will never become a (more) whitespace sensitive language, just as C# will never get rid of semicolons.  The level of effort for such projects is enormous and the value of doing so would be dubious at best.  The .NET runtime was designed to be a polyglot runtime and there are many languages available for it, there's no reason to make C# feel like every other language.

**Reply by TahirAhmadov** [reply]
No argument, I'm fully in agreement; my point is that I've seen this asked a few times already, and if we ever get an "FAQ", this should be a line item there.
Regarding making C# a better option to those beginner and non-software-developer users, I don't mean by changing its syntax; rather through tooling, outreach, education, collaborations, etc.

> The .NET runtime was designed to be a polyglot runtime and there are many languages available for it, there's no reason to make C# feel like every other language.

Actually, that reminded me of this: https://ironpython.net/
Those who want the power of .NET BCL and interoperability with other .NET code, such as stuff written in C#, while preferring the Python syntax and style, it's an excellent option to explore. @Developer-Alexander

**Reply by Developer-Alexander** [reply]
I find that questionable. For me, the underlying question is what C# actually wants to be.
Does C# want to be a tool that meets users' needs? Or does C# want to be an end in itself. (Regardless of which of these applies to indented blocks)
And I think you really have to ask yourself that question. What would the majority of users think about this feature? Otherwise you run the risk that the language extensions you make will only be of use to a few in certain borderline cases.

I have a bit of the impression that the discussion was already lost when I mentioned Python in the title. From then on, the discussion no longer had any chance of success.

I also have the impression that there is a lot of cherry picking going on in the arguments, so I can't quite see the common thread.
The discussion about optional semicolons is a very good example of this. (Although I'm not a fanboy of optional semicolons now.)
People like to threaten with ambiguities in the code, but this already exists elsewhere in the code. For example, when one of the compilers indicates that it could not automatically recognize the type of generics. And that in turn is a very elegant solution to such problems.

And I would like to mention one important aspect that I'm not sure if everyone is always aware of. Special characters in particular are sometimes much more difficult to access on keyboards in other languages, and this may be one of the reasons why people prefer programming languages ​​that get by with significantly fewer of these characters.

Last but not least, I find the argument inadmissible when it is argued that a new feature is not needed because the .Net ecosystem already has various great languages. That's not even an argument.

**Reply by HaloFour** [reply]
@Developer-Alexander 

Your proposal is very specifically about making a change to the language to meet a stylistic preference.  There is no goal for the language to meet the stylistic preferences of all users who are coming from other languages.

If you don't like the rest of the arguments that's your business, but there's zero chance that C# will accept this proposal, as has been expressed by a few members of the language design team.  Even if it was _easy_ it's not worth doing because it solves exactly _zero_ problems.

**Reply by TahirAhmadov** [reply]
@Developer-Alexander the problem with modifying C# like that isn't necessarily that people dislike Python per se, rather that the proposed change is (as explained by others) a) extremely expensive, b) has no value, c) actually has downsides of its own.

To your question about direction of C# in general - yes the team does aim to serve users' needs, but the problem is that your proposal is absolute rock bottom when it comes to the priorities (even outside of the issues I listed above). People do request this from time to time, but MANY more need other stuff.

Regarding .NET languages, I'm not sure why you find it "inadmissible". That's the entire point of separating the runtime and the languages, like it is with .NET and JVM. There are millions of software developers and obviously, it's impossible to make one language which suits everybody's tastes and preferences. That's why anybody can create their own language, or contribute to an existing one (like IronPython), and still get all the benefits of .NET.

**Reply by DavidArno** [reply]
> Actually, that reminded me of this: https://ironpython.net/

Slight problem that the current beta release is only compatible with Python 3.4 code, which was released eight years ago. IronPython is an interesting project but isn't really suitable for serious use.

> ... people start asking questions like, "why isn't your language like Python?" - it became "native tongue" for too many; a good reminder to reinforce efforts to make C# an attractive option to those users.

That seems a rather arrogant assertion to me. "People learning Python is bad, we must find ways of ensuring they learn C# instead". Python syntax (in fact Python itself) has pros and cons. There's no right answer here and it certainly isn't the case that Python is just a "beginners' language".

**Reply by CyrusNajmabadi** [reply]
There are many different stylistic approaches commonly accepted by programming languages.  Some languages even allow the user to choose between a few of them.

C# is not currently one of those languages, and currently has no aim to be.  We've chosen a syntax intentionally and we believe there is greater value to the ecosystem picking and sticking with one style versus allowing any other styles as a user choice.

**Reply by TahirAhmadov** [reply]
> Slight problem that the current beta release is only compatible with Python 3.4 code, which was released eight years ago. IronPython is an interesting project but isn't really suitable for serious use.

Sure, but the only way to ensure that language is developed is to use it and contribute to it. Otherwise, it's a catch-22.

> That seems a rather arrogant assertion to me. "People learning Python is bad, we must find ways of ensuring they learn C# instead". Python syntax (in fact Python itself) has pros and cons. There's no right answer here and it certainly isn't the case that Python is just a "beginners' language".

The indentation style of Python, for all of its downsides, is nowhere near the top when it comes to the bigger problems with it, when compared to C#; I know others may have other priorities, but I think objectively C# is a better language all around.
And I didn't mean to say Python is for beginners only; it has a massive user base of professionals, including software pros, as well as scientists, etc. My contention is that, its massive popularity prevents a lot of people from even trying to use C#/.NET, and leaving aside the reduced usage of C# as a result, but it also doesn't allow those users from even making a choice - I'm certain many (but not all, obviously) would prefer C#, given the chance to try both it and Python.

**Comment by Xyncgas** [comment]
On the other hand I would like brackets on F#
